app/vmctl/opentsdb.go

```diff
@@ -0,0 +1,168 @@
+		m, err := op.oc.FindMetrics(q)
+		if err != nil {
+			return fmt.Errorf("metric discovery failed for %q: %s", q, err)
+		}
+		metrics = append(metrics, m...)
+	}
+	if len(metrics) < 1 {
+		return fmt.Errorf("found no timeseries to import with filters %q", op.oc.Filters)
+	}
+
+	question := fmt.Sprintf("Found %d metrics to import. Continue?", len(metrics))
+	if !silent && !prompt(question) {
+		return nil
+	}
+	op.im.ResetStats()
+	startTime := time.Now().Unix()
+	queryRanges := 0
+	// pre-calculate the number of query ranges we'll be processing
+	for _, rt := range op.oc.Retentions {
+		queryRanges += len(rt.QueryRanges)
+	}
+	for _, metric := range metrics {
+		log.Println(fmt.Sprintf("Starting work on %s", metric))
+		serieslist, err := op.oc.FindSeries(metric)
+		//log.Println(fmt.Sprintf("Found %d series for %s", len(serieslist), metric))
+		if err != nil {
+			return fmt.Errorf("couldn't retrieve series list for %s : %s", metric, err)
+		}
+		/*
+			Create channels for collecting/processing series and errors
+			We'll create them per metric to reduce pressure against OpenTSDB
+
+			Limit the size of seriesCh so we can't get too far ahead of actual processing
+		*/
+		seriesCh := make(chan queryObj, op.otsdbcc)
+		errCh := make(chan error)
+		// we're going to make serieslist * queryRanges queries, so we should represent that in the progress bar
+		bar := pb.StartNew(len(serieslist) * queryRanges)
+		var wg sync.WaitGroup
+		wg.Add(op.otsdbcc)
+		for i := 0; i < op.otsdbcc; i++ {
+			go func() {
+				defer wg.Done()
+				for s := range seriesCh {
+					if err := op.do(s); err != nil {
+						errCh <- fmt.Errorf("couldn't retrieve series for %s : %s", metric, err)
+						return
+					}
+					bar.Increment()
+				}
+			}()
+		}
+		/*
+			Loop through all series for this metric, processing all retentions and time ranges
+			requested. This loop is our primary "collect data from OpenTSDB loop" and should
+			be async, sending data to VictoriaMetrics over time.
+
+			The idea with having the select at the inner-most loop is to ensure quick
+			short-circuiting on error.
+		*/
+		for _, series := range serieslist {
+			for _, rt := range op.oc.Retentions {
+				for _, tr := range rt.QueryRanges {
+					select {
+					case otsdbErr := <-errCh:
+						return fmt.Errorf("opentsdb error: %s", otsdbErr)
+					case vmErr := <-op.im.Errors():
+						return fmt.Errorf("Import process failed: \n%s", wrapErr(vmErr))
+					case seriesCh <- queryObj{
+						Tr: tr, StartTime: startTime,
+						Series: series, Rt: opentsdb.RetentionMeta{
+							FirstOrder: rt.FirstOrder, SecondOrder: rt.SecondOrder, AggTime: rt.AggTime}}:
+					}
+				}
+			}
+		}
+		// Drain channels per metric
+		close(seriesCh)
+		wg.Wait()
+		close(errCh)
+		// check for any lingering errors on the query side
+		for otsdbErr := range errCh {
+			return fmt.Errorf("Import process failed: \n%s", otsdbErr)
+		}
+		bar.Finish()
+		log.Print(op.im.Stats())
```

Comment:
should we reset stats on each iteration?

Reply:
Well...I wasn't sure. I thought keeping the import stats across the entire run, rather than only for each metric, would be nicer. But if you prefer one way over the other...

Reply:
I wasn't sure the right move here. I thought having a running count of data would be interesting as we process metrics. But if only showing what happened per metric makes more sense, it's certainly easy to set that instead.

Reply:
I don't have strong opinion here. I guess, it is up to you because I'm not familiar with opentsdb at all.